File: smarttrade-web/src/features/boleta-inteligente/data/repositories/instrumentRepositoryImpl.ts
import { NotFoundFailure } from '@/core/error/failure'
import { err, ok, type Result } from '@/core/result/result'
import type { Instrument } from '../../domain/entities/instrument'
import type { InstrumentRepository } from '../../domain/repositories/instrumentRepository'
import { seedInstruments } from '../datasources/instrumentLocalDataSource'

export class InstrumentRepositoryImpl implements InstrumentRepository {
  async getAvailableInstruments(): Promise<Result<Instrument[]>> {
    return ok(seedInstruments)
  }

  async getInstrumentBySymbol(symbol: string): Promise<Result<Instrument>> {
    const instrument = seedInstruments.find((i) => i.symbol === symbol)
    if (!instrument) {
      return err(new NotFoundFailure(`Instrumento "${symbol}" não encontrado.`))
    }
    return ok(instrument)
  }
}

export const instrumentRepository: InstrumentRepository = new InstrumentRepositoryImpl()

File: smarttrade_app/lib/features/boleta_inteligente/data/repositories/instrument_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:smarttrade_app/core/error/failures.dart';
import 'package:smarttrade_app/features/boleta_inteligente/data/datasources/instrument_local_datasource.dart';
import 'package:smarttrade_app/features/boleta_inteligente/domain/entities/instrument.dart';
import 'package:smarttrade_app/features/boleta_inteligente/domain/repositories/instrument_repository.dart';

class InstrumentRepositoryImpl implements InstrumentRepository {
  const InstrumentRepositoryImpl();

  @override
  Future<Either<Failure, List<Instrument>>> getAvailableInstruments() async {
    return Right(InstrumentLocalDataSource.seedInstruments);
  }

  @override
  Future<Either<Failure, Instrument>> getInstrumentBySymbol(
    String symbol,
  ) async {
    for (final instrument in InstrumentLocalDataSource.seedInstruments) {
      if (instrument.symbol == symbol) {
        return Right(instrument);
      }
    }
    return Left(NotFoundFailure('Instrumento "$symbol" não encontrado.'));
  }
}
